### PR TITLE
[quant][pt2e][xnnpack] Add support for QAT dynamic quantization for linear in XNNPACKQuantizer

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113288

Summary:
FX graph mode quant workflow and also pt2e flow relies on the `is_dynamic` flag in observer/quantizationspec to
convert an observer to dynamic quantization patterns (choose_qparams -> q -> dq), this PR added is_dynamic flag
for all observers so that it's possible to convert these observers to the pattern.

However, this dynamic quantization pattern (choose_qparams -> q -> dq) is actually only valid for MovingAverageObserver(averaging_constant=1)
for the computation before convert and after convert to match in the context of QAT. So we'll have some sanity
checks in other observers to make sure the is_dynamic is False.

Test Plan:
python test/test_quantization.py TestXNNPACKQuantizer.test_qat_dynamic_linear

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D51124725](https://our.internmc.facebook.com/intern/diff/D51124725)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51124725).

File: torch/ao/quantization/observer.py [hunk 1]
```diff
@@ -133,11 +133,14 @@ class ObserverBase(ABC, nn.Module):
...
                reference model spec.
+        is_dynamic: indicator for whether the observer is a placeholder for dynamic quantization
+        or static quantization
     """
 
-    def __init__(self, dtype):
+    def __init__(self, dtype, is_dynamic=False):
         super().__init__()
         self.dtype = dtype
+        self.is_dynamic = False
```
Comment (kimishpatel) [inline torch/ao/quantization/observer.py:143]: Did you mean to use `is_dynamic` from the arg?

File: torch/ao/quantization/fake_quantize.py [hunk 2]
```diff
@@ -147,7 +150,7 @@ class FakeQuantize(FakeQuantizeBase):
     scale: torch.Tensor
     zero_point: torch.Tensor
 
-    def __init__(self, observer=MovingAverageMinMaxObserver, quant_min=None, quant_max=None, **observer_kwargs):
+    def __init__(self, observer=MovingAverageMinMaxObserver, quant_min=None, quant_max=None, is_dynamic=False, **observer_kwargs):
```
Comment (kimishpatel) [inline torch/ao/quantization/fake_quantize.py:153]: I also dont understand what it means to have is_dynamic for moving average min/max observer. Doesnt feel right

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Can you add context for the need of this change. not clear to me

File: torch/ao/quantization/observer.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/observer.py:143]: oh right

File: torch/ao/quantization/fake_quantize.py (hunk 2 above)
Comment (kimishpatel) [reply torch/ao/quantization/fake_quantize.py:153]: It feel like only MinMaxObserver/HistogramObserver can have is_dynamic. Anything that requires larger window of observeration cannot be called dynamically quantized? So thats my question.

Comment (kimishpatel) [reply torch/ao/quantization/fake_quantize.py:153]: My question is this: For dynamic quant, can we not force that the observer must be Histogram or MinMax, not moving average. Not sure if thats what you answered above

Comment (jerryzh168) [reply torch/ao/quantization/fake_quantize.py:153]: right now we force it must be MovingAverageMinMaxObserver with averaging_constant=1, this is the observer+configuration that mimics the computation of dynamic quant exactly in QAT and after convert (which is choose_qparams on a single input + q + dq). normal minmax observer or histogram observer does not have this behavior so they do not have corresponding dynamic quant simulation/conversion.

Comment (jerryzh168) [reply torch/ao/quantization/fake_quantize.py:153]: minmax/histogram works for static QAT.

this is just talking about dynamic quantization (QAT/PTQ), so currently in convert, the pattern we produce for dynamic quantization is `choose_qparams -> q -> dq` and the computation is equivalent to AveragingMinMaxObserver with averaging_constant = 1.. 

all other computations does not have a corresponding dynamic quant pattern, e.g. histogram observer/minmax observer or averaging minmax observer with different averaging_constant all requires observing multiple inputs and maintaining some states, this is different from dynamic quantization which does not have any state, that's why MovingAverageMinMaxObserver with averaging_constant=1 (meaning no state) is the only thing that corresponds to the current implementation of dynamic quantization.

**Review by kimishpatel — APPROVED** [review]

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51124725).

**Pull request closed without merging**